File: AgentStateTesting/python-tests/basic-sdk-test.py
#!/usr/bin/env python3
"""
Basic AgentState SDK Testing
============================

Tests basic functionality of the AgentState Python SDK including:
- Connection and health checks
- CRUD operations
- Error handling
- Authentication
"""

import os
import time
import pytest
from typing import Dict, Any
from agentstate import AgentStateClient
from colorama import init, Fore, Style

# Initialize colorama for colored output
init()

# Configuration
AGENTSTATE_URL = os.getenv('AGENTSTATE_URL', 'http://localhost:8080')
AGENTSTATE_API_KEY = os.getenv('AGENTSTATE_API_KEY')  # Optional
import uuid
NAMESPACE = f'testing-{uuid.uuid4().hex[:8]}'

def print_success(message: str):
    print(f"{Fore.GREEN}✅ {message}{Style.RESET_ALL}")

def print_error(message: str):
    print(f"{Fore.RED}❌ {message}{Style.RESET_ALL}")

def print_info(message: str):
    print(f"{Fore.BLUE}ℹ️  {message}{Style.RESET_ALL}")

class TestAgentStateSDK:
    
    def setup_method(self):
        """Set up test client"""
        self.client = AgentStateClient(
            base_url=AGENTSTATE_URL,
            namespace=NAMESPACE,
            api_key=AGENTSTATE_API_KEY
        )
        self.test_agents = []  # Track created agents for cleanup
        
    def teardown_method(self):
        """Clean up test agents"""
        for agent_id in self.test_agents:
            try:
                self.client.delete_agent(agent_id)
                print_info(f"Cleaned up agent: {agent_id}")
            except Exception as e:
                print_error(f"Failed to clean up agent {agent_id}: {e}")
    
    def test_health_check(self):
        """Test server health check"""
        print_info("Testing health check...")
        is_healthy = self.client.health_check()
        assert is_healthy, "Server should be healthy"
        print_success("Health check passed")
    
    def test_create_agent(self):
        """Test creating an agent"""
        print_info("Testing agent creation...")
        
        agent = self.client.create_agent(
            agent_type="test-bot",
            body={
                "name": "TestBot",
                "status": "active",
                "created_at": time.time()
            },
            tags={
                "test": "true",
                "framework": "sdk-test",
                "environment": "testing"
            }
        )
        
        assert agent['type'] == "test-bot"
        assert agent['body']['name'] == "TestBot"
        assert agent['tags']['test'] == "true"
        assert 'id' in agent
        assert 'commit_seq' in agent
        assert 'ts' in agent
        
        self.test_agents.append(agent['id'])
        print_success(f"Created agent: {agent['id']}")
        return agent
    
    def test_get_agent(self):
        """Test retrieving an agent by ID"""
        print_info("Testing agent retrieval...")
        
        # First create an agent
        created_agent = self.test_create_agent()
        
        # Then retrieve it
        retrieved_agent = self.client.get_agent(created_agent['id'])
        
        assert retrieved_agent['id'] == created_agent['id']
        assert retrieved_agent['type'] == created_agent['type']
        assert retrieved_agent['body'] == created_agent['body']
        assert retrieved_agent['tags'] == created_agent['tags']
        
        print_success("Agent retrieval successful")
    
    def test_update_agent(self):
        """Test updating an existing agent"""
        print_info("Testing agent update...")
        
        # Create initial agent
        agent = self.test_create_agent()
        original_seq = agent['commit_seq']
        
        # Update the agent
        updated_agent = self.client.create_agent(
            agent_type="test-bot",
            body={
                "name": "UpdatedTestBot",
                "status": "busy",
                "updated_at": time.time()
            },
            tags={
                "test": "true",
                "framework": "sdk-test",
                "environment": "testing",
                "updated": "true"
            },
            agent_id=agent['id']  # Update existing
        )
        
        assert updated_agent['id'] == agent['id']
        assert updated_agent['body']['name'] == "UpdatedTestBot"
        assert updated_agent['body']['status'] == "busy"
        assert updated_agent['tags']['updated'] == "true"
        assert updated_agent['commit_seq'] > original_seq
        
        print_success("Agent update successful")
    
    def test_query_agents(self):
        """Test querying agents by tags"""
        print_info("Testing agent querying...")
        
        # Create multiple test agents
        agents = []
        for i in range(3):
            priority = "high" if i % 2 == 0 else "low"
            agent = self.client.create_agent(
                agent_type="query-test",
                body={
                    "name": f"QueryBot{i}",
                    "index": i
                },
                tags={
                    "test": "true",
                    "batch": "query-test", 
                    "priority": priority
                }
            )
            print_info(f"Created agent {i} with priority: {priority}")
            agents.append(agent)
            self.test_agents.append(agent['id'])
        
        # Query by batch tag
        batch_results = self.client.query_agents({"batch": "query-test"})
        if len(batch_results) != 3:
            print_error(f"Expected 3 agents, got {len(batch_results)}: {[r['id'] for r in batch_results]}")
        assert len(batch_results) == 3
        print_success(f"Found {len(batch_results)} agents in batch")
        
        # Query by priority tag only (since AND filtering may not be supported)
        high_priority = self.client.query_agents({"priority": "high"})
        high_count = len([a for a in high_priority if a['tags'].get('batch') == 'query-test'])
        if high_count < 2:
            print_error(f"Expected at least 2 high priority agents, got {high_count}")
        # Just verify we can query by a single tag successfully
        assert len(high_priority) >= 2
        print_success(f"Found {len(high_priority)} high priority agents (at least 2 from our batch)")
    
    def test_delete_agent(self):
        """Test deleting an agent"""
        print_info("Testing agent deletion...")
        
        # Create an agent
        agent = self.test_create_agent()
        agent_id = agent['id']
        
        # Verify it exists
        retrieved = self.client.get_agent(agent_id)
        assert retrieved['id'] == agent_id
        
        # Delete it
        self.client.delete_agent(agent_id)
        
        # Verify it's gone
        try:
            self.client.get_agent(agent_id)
            assert False, "Agent should have been deleted"
        except Exception as e:
            assert "404" in str(e) or "not found" in str(e).lower()
            
        print_success("Agent deletion successful")
        
        # Remove from cleanup list since it's already deleted
        if agent_id in self.test_agents:
            self.test_agents.remove(agent_id)
    
    def test_error_handling(self):
        """Test error handling for invalid operations"""
        print_info("Testing error handling...")
        
        # Test getting non-existent agent
        try:
            self.client.get_agent("non-existent-id")
            assert False, "Should have raised an exception"
        except Exception as e:
            assert "404" in str(e) or "not found" in str(e).lower()
            print_success("404 error handled correctly")
        
        # Test deleting non-existent agent
        try:
            self.client.delete_agent("non-existent-id")
            assert False, "Should have raised an exception"
        except Exception as e:
            assert "404" in str(e) or "not found" in str(e).lower()
            print_success("Delete 404 error handled correctly")

def run_manual_tests():
    """Run tests manually without pytest"""
    print(f"{Fore.CYAN}🧪 AgentState SDK Basic Testing{Style.RESET_ALL}")
    print(f"Server: {AGENTSTATE_URL}")
    print(f"Namespace: {NAMESPACE}")
    print(f"API Key: {'Set' if AGENTSTATE_API_KEY else 'Not set'}")
    print("-" * 50)
    
    test_instance = TestAgentStateSDK()
    
    tests = [
        test_instance.test_health_check,
        test_instance.test_create_agent,
        test_instance.test_get_agent,
        test_instance.test_update_agent,
        test_instance.test_query_agents,
        test_instance.test_delete_agent,
        test_instance.test_error_handling,
    ]
    
    passed = 0
    failed = 0
    
    for test in tests:
        try:
            test_instance.setup_method()
            test()
            test_instance.teardown_method()
            passed += 1
        except Exception as e:
            print_error(f"Test {test.__name__} failed: {e}")
            import traceback
            traceback.print_exc()
            failed += 1
    
    print("-" * 50)
    print(f"{Fore.CYAN}Results: {passed} passed, {failed} failed{Style.RESET_ALL}")
    
    if failed == 0:
        print(f"{Fore.GREEN}🎉 All tests passed!{Style.RESET_ALL}")
    else:
        print(f"{Fore.RED}💥 {failed} test(s) failed{Style.RESET_ALL}")

if __name__ == "__main__":
    run_manual_tests()

File: AgentStateTesting/nodejs-tests/basic-sdk-test.js
#!/usr/bin/env node
/**
 * Basic AgentState SDK Testing (Node.js/TypeScript)
 * ==================================================
 * 
 * Tests basic functionality of the AgentState TypeScript/JavaScript SDK including:
 * - Connection and health checks
 * - CRUD operations  
 * - Error handling
 * - Authentication
 */

import 'dotenv/config';
import { AgentStateClient } from 'agentstate';
import colors from 'colors';

// Configuration
const AGENTSTATE_URL = process.env.AGENTSTATE_URL || 'http://localhost:8080';
const AGENTSTATE_API_KEY = process.env.AGENTSTATE_API_KEY; // Optional
const NAMESPACE = 'testing';

// Helper functions for colored output
const success = (msg) => console.log(colors.green(`✅ ${msg}`));
const error = (msg) => console.log(colors.red(`❌ ${msg}`));
const info = (msg) => console.log(colors.blue(`ℹ️  ${msg}`));

class AgentStateSDKTester {
  constructor() {
    this.client = new AgentStateClient(AGENTSTATE_URL, NAMESPACE, AGENTSTATE_API_KEY);
    this.testAgents = []; // Track created agents for cleanup
  }

  async cleanup() {
    /**Clean up test agents*/
    for (const agentId of this.testAgents) {
      try {
        await this.client.deleteAgent(agentId);
        info(`Cleaned up agent: ${agentId}`);
      } catch (e) {
        error(`Failed to clean up agent ${agentId}: ${e.message}`);
      }
    }
    this.testAgents = [];
  }

  async testHealthCheck() {
    /**Test server health check*/
    info('Testing health check...');
    const isHealthy = await this.client.healthCheck();
    if (!isHealthy) {
      throw new Error('Server should be healthy');
    }
    success('Health check passed');
  }

  async testCreateAgent() {
    /**Test creating an agent*/
    info('Testing agent creation...');
    
    const agent = await this.client.createAgent(
      'test-bot',
      {
        name: 'TestBot',
        status: 'active',
        createdAt: Date.now()
      },
      {
        test: 'true',
        framework: 'sdk-test',
        environment: 'testing'
      }
    );
    
    if (agent.type !== 'test-bot') {
      throw new Error(`Expected type 'test-bot', got '${agent.type}'`);
    }
    if (agent.body.name !== 'TestBot') {
      throw new Error(`Expected name 'TestBot', got '${agent.body.name}'`);
    }
    if (agent.tags.test !== 'true') {
      throw new Error(`Expected tag test='true', got '${agent.tags.test}'`);
    }
    if (!agent.id || !agent.commit_seq || !agent.commit_ts) {
      throw new Error('Agent missing required fields: id, commit_seq, commit_ts');
    }
    
    this.testAgents.push(agent.id);
    success(`Created agent: ${agent.id}`);
    return agent;
  }

  async testGetAgent() {
    /**Test retrieving an agent by ID*/
    info('Testing agent retrieval...');
    
    // First create an agent
    const createdAgent = await this.testCreateAgent();
    
    // Then retrieve it
    const retrievedAgent = await this.client.getAgent(createdAgent.id);
    
    if (retrievedAgent.id !== createdAgent.id) {
      throw new Error('Retrieved agent ID does not match created agent ID');
    }
    if (JSON.stringify(retrievedAgent.body) !== JSON.stringify(createdAgent.body)) {
      throw new Error('Retrieved agent body does not match created agent body');
    }
    if (JSON.stringify(retrievedAgent.tags) !== JSON.stringify(createdAgent.tags)) {
      throw new Error('Retrieved agent tags do not match created agent tags');
    }
    
    success('Agent retrieval successful');
  }

  async testUpdateAgent() {
    /**Test updating an existing agent*/
    info('Testing agent update...');
    
    // Create initial agent
    const agent = await this.testCreateAgent();
    const originalSeq = agent.commit_seq;
    
    // Update the agent
    const updatedAgent = await this.client.createAgent(
      'test-bot',
      {
        name: 'UpdatedTestBot',
        status: 'busy',
        updatedAt: Date.now()
      },
      {
        test: 'true',
        framework: 'sdk-test',
        environment: 'testing',
        updated: 'true'
      },
      agent.id // Update existing
    );
    
    if (updatedAgent.id !== agent.id) {
      throw new Error('Updated agent ID does not match original agent ID');
    }
    if (updatedAgent.body.name !== 'UpdatedTestBot') {
      throw new Error(`Expected updated name 'UpdatedTestBot', got '${updatedAgent.body.name}'`);
    }
    if (updatedAgent.body.status !== 'busy') {
      throw new Error(`Expected updated status 'busy', got '${updatedAgent.body.status}'`);
    }
    if (updatedAgent.tags.updated !== 'true') {
      throw new Error(`Expected updated tag updated='true', got '${updatedAgent.tags.updated}'`);
    }
    if (updatedAgent.commit_seq <= originalSeq) {
      throw new Error('Updated agent should have higher commit_seq');
    }
    
    success('Agent update successful');
  }

  async testQueryAgents() {
    /**Test querying agents by tags*/
    info('Testing agent querying...');
    
    // Create multiple test agents
    const agents = [];
    for (let i = 0; i < 3; i++) {
      const agent = await this.client.createAgent(
        'query-test',
        {
          name: `QueryBot${i}`,
          index: i
        },
        {
          test: 'true',
          batch: 'query-test',
          priority: i % 2 === 0 ? 'high' : 'low'
        }
      );
      agents.push(agent);
      this.testAgents.push(agent.id);
    }
    
    // Query by batch tag
    const batchResults = await this.client.queryAgents({ batch: 'query-test' });
    if (batchResults.length !== 3) {
      throw new Error(`Expected 3 agents in batch, got ${batchResults.length}`);
    }
    success(`Found ${batchResults.length} agents in batch`);
    
    // Query by priority tag
    const highPriority = await this.client.queryAgents({
      batch: 'query-test',
      priority: 'high'
    });
    if (highPriority.length !== 2) { // indices 0 and 2
      throw new Error(`Expected 2 high priority agents, got ${highPriority.length}`);
    }
    success(`Found ${highPriority.length} high priority agents`);
  }

  async testDeleteAgent() {
    /**Test deleting an agent*/
    info('Testing agent deletion...');
    
    // Create an agent
    const agent = await this.testCreateAgent();
    const agentId = agent.id;
    
    // Verify it exists
    const retrieved = await this.client.getAgent(agentId);
    if (retrieved.id !== agentId) {
      throw new Error('Agent should exist before deletion');
    }
    
    // Delete it
    await this.client.deleteAgent(agentId);
    
    // Verify it's gone
    try {
      await this.client.getAgent(agentId);
      throw new Error('Agent should have been deleted');
    } catch (e) {
      if (!e.message.includes('404') && !e.message.toLowerCase().includes('not found')) {
        throw new Error(`Expected 404 error, got: ${e.message}`);
      }
    }
    
    success('Agent deletion successful');
    
    // Remove from cleanup list since it's already deleted
    const index = this.testAgents.indexOf(agentId);
    if (index > -1) {
      this.testAgents.splice(index, 1);
    }
  }

  async testErrorHandling() {
    /**Test error handling for invalid operations*/
    info('Testing error handling...');
    
    // Test getting non-existent agent
    try {
      await this.client.getAgent('non-existent-id');
      throw new Error('Should have raised an exception');
    } catch (e) {
      if (!e.message.includes('404') && !e.message.toLowerCase().includes('not found')) {
        throw new Error(`Expected 404 error, got: ${e.message}`);
      }
      success('404 error handled correctly');
    }
    
    // Test deleting non-existent agent
    try {
      await this.client.deleteAgent('non-existent-id');
      throw new Error('Should have raised an exception');
    } catch (e) {
      if (!e.message.includes('404') && !e.message.toLowerCase().includes('not found')) {
        throw new Error(`Expected 404 error, got: ${e.message}`);
      }
      success('Delete 404 error handled correctly');
    }
  }

  async runAllTests() {
    /**Run all tests*/
    console.log(colors.cyan('🧪 AgentState SDK Basic Testing'));
    console.log(`Server: ${AGENTSTATE_URL}`);
    console.log(`Namespace: ${NAMESPACE}`);
    console.log(`API Key: ${AGENTSTATE_API_KEY ? 'Set' : 'Not set'}`);
    console.log('-'.repeat(50));
    
    const tests = [
      this.testHealthCheck,
      this.testCreateAgent, 
      this.testGetAgent,
      this.testUpdateAgent,
      this.testQueryAgents,
      this.testDeleteAgent,
      this.testErrorHandling,
    ];
    
    let passed = 0;
    let failed = 0;
    
    for (const test of tests) {
      try {
        await test.call(this);
        await this.cleanup();
        passed++;
      } catch (e) {
        error(`Test ${test.name} failed: ${e.message}`);
        await this.cleanup();
        failed++;
      }
    }
    
    console.log('-'.repeat(50));
    console.log(colors.cyan(`Results: ${passed} passed, ${failed} failed`));
    
    if (failed === 0) {
      console.log(colors.green('🎉 All tests passed!'));
    } else {
      console.log(colors.red(`💥 ${failed} test(s) failed`));
      process.exit(1);
    }
  }
}

// Run tests if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  const tester = new AgentStateSDKTester();
  tester.runAllTests().catch(e => {
    error(`Test suite failed: ${e.message}`);
    process.exit(1);
  });
}